dockertest/dockercmd.py

```diff
@@ -334,6 +334,49 @@ def execute(self, stdin=None):
                                          stdin=stdin, close_fds=True)
         return self.cmdresult
 
+    def wait_for_ready(self, timeout=None, timestep=0.2):
+        """
+        Monitor the output of a container (including docker logs, in
+        case stdout is detached), waiting for the string 'READY'. Return
+        as soon as we see it. If we don't, throw a meaningful exception.
+
+        :raises DockerExecError: on timeout.
+        """
+        end_time = time.time() + timeout
+        while time.time() <= end_time:
+            time.sleep(timestep)
+            if 'READY' in self.stdout:
+                return
+            # Also check docker logs
+            cid = self.container_id
+            if cid is not None:
+                logs = DockerCmd(self.subtest, 'logs', [self.container_id])
+                logs.execute()
+                if 'READY' in logs.stdout:
+                    return
+
+        # Never saw READY. Did container exit? If so, help user understand why
+        if self.done:
+            msg = "Container exited before READY"
+            if self.exit_status == 0:
+                msg += " (normal exit status)"
+            else:
+                msg += "; exit status = %d" % self.exit_status
+            stderr = self.stderr
+            if stderr:
+                msg += "; stderr='%s'" % stderr
+            raise DockerExecError(msg)
+
+        # Container still running. Must be a timeout.
+        msg = "Timed out waiting for container READY"
+        stdout = self.stdout
+        if stdout:
+            # Juuuuuuust in case it appeared since we last checked
+            if 'READY' in stdout:
+                return
```

Comment:
The reason for this is that the `self.stdout` grabbed in line 372 is not the same as the one checked earlier in the function. This one is being displayed to the user as part of an error message, and I've seen (on two occasions) moronic error messages that tell me "the expected string 'X' is not in 'XYZ'", and it's obvious that X **is** in XYZ, but the string must've appeared in the small window between checking and producing the error message.

I know it's duplication, and it's horrible, but without a way to preserve the last-checked stdout I think this is necessary. The alternative is to completely remove the quoted stdout, just telling the reader "tough luck, figure it out on your own". And I think that's worse than duplication. (Which takes a lot for me to say.)